uint128/uint128.go

```diff
@@ -0,0 +1,110 @@
+package uint128
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"sync"
+)
+
+// *Int[1] = high bits, more significant bits, upper 64 bits of a 128-bit value - ie [64:128]
+// *Int[0] = low bits, less significant bits, lower 64 bits of a 128-bit value - ie [0:64]
+// for example, if we have a 128-bit hex value 0x6e58133b38301a6cdfa34ca991c4ba39
+// *Int[0] (lower bits) should be 0xdfa34ca991c4ba39 or dfa34ca991c4ba39
+// *Int[i] (upper bits) should be 0x6e58133b38301a6c or 6e58133b38301a6c
+// an example of a call to this type can be : uint128.Int([]uint64{lo, high})
+type Int [2]uint64
+
+const (
+	Base16 = 16
+	Bytes  = 16
+	Bits   = 128
+)
+
+var bigIntPool = sync.Pool{
+	New: func() interface{} {
+		return new(big.Int)
+	},
+}
+
+func (i *Int) SetBigInt(v *big.Int) *Int {
+	words := v.Bits()
+
+	if len(words) > 0 {
+		switch {
+		case len(words)%2 == 0:
+			i[0] = uint64(words[0])
+			i[1] = uint64(words[1])
+		case len(words)%2 == 1:
+			i[0] = uint64(words[0])
+		}
+	}
+
```

Comment:
We should align with `felt`'s behavior, which performs the modular reduction when `v` cannot fit in a felt. We can take a lot of inspiration from `felt.SetBigInt` and `felt.setBigInt`.